db: allow adding racks with RF-rack-valid keyspaces

wmitros opened the pull request:
In this patch we lift the restriction of refusing to add new racks to datacenters while RF-rack-valid keyspaces exist. Before this patch, if we enabled the RF-rack-valid-keyspaces option and created a keyspace, no new racks could be added, as these keyspaces would be no longer RF-rack-valid. The option affects all keyspaces with
 tablets, so in practice, if we had any data in the DC, we could no longer add or remove any racks.
This patch doesn't enable all configurations of keyspace RFs and racks, but it enables some extra configurations, not possible before:
* adding a new rack
* removing a rack, as long as no RF-rack-valid keyspace (with RF>1) is replicated on it
* increasing the RF of a keyspace, if it makes it RF-rack-valid

After a new rack is added, the usage of the existing RF-rack-valid keyspaces is limited - we can no longer create new tables in it until the RF is increased to match the number of racks.

Allowing adding new racks was accomplished by the following steps:
* on node startup, instead of checking whether all keyspaces are replicated on all racks, for each keyspace we now check if all tablets in all tables in this keyspace are replicated on the same racks
* the load balancer is adjusted to reject any cross-rack migrations of RF-rack-valid keyspaces, instead of only discouraging them
* when creating a table in a keyspace, to avoid allocating its tablets on too many racks, we now check whether the keyspace is RF-rack-valid

With these steps we make sure that at any point in time, every table in a (originally) RF-rack-valid keyspace with a given RF will be replicated on specific RF racks.

After this patch, we still can't decrease the RF of a keyspace, increase it to a number lower than the number of racks and we can't remove a rack that already contains some data from RF-rack-valid keyspaces, unless we drop these keyspaces before. These issues will need to be fixed in an additional, more comprehensive patch.

Fixes: https://github.com/scylladb/scylladb/issues/23426

wmitros commented:
> #### Failed Tests (2/44912):
> * [test_replace_with_same_ip_twice.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16562/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.release.1)
> * [test_replace_with_same_ip_twice.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16562/testReport/junit/cluster/test_replace_with_same_ip_twice/Tests___Unit_Tests___test_replace_with_same_ip_twice_release_1_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice.release.1)


https://github.com/scylladb/scylladb/issues/23553

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16566/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16566/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16566/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_multidc
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7695/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4553/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15746/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16566/)
#### Build Details:
- Duration: 7 hr 22 min
- Builder: spider5.cloudius-systems.com

piodul commented on cql3/statements/create_table_statement.cc:
```diff
@@ -72,6 +72,30 @@ std::vector<column_definition> create_table_statement::get_columns() const
 create_table_statement::prepare_schema_mutations(query_processor& qp, const query_options&, api::timestamp_type ts) const {
+    const auto tmptr = qp.proxy().get_token_metadata_ptr();
+    const auto& cfg = qp.db().get_config();
+    // FIXME: After adding a new rack, the keyspace is not RF-rack-valid anymore,
```
Currently, the definition of the `--rf-rack-valid-keyspaces` option is like this:

```
    , rf_rack_valid_keyspaces(this, "rf_rack_valid_keyspaces", liveness::MustRestart, value_status::Used, false,
        "Enforce RF-rack-valid keyspaces. Additionally, if there are existing RF-rack-invalid "
        "keyspaces, attempting to start a node with this option ON will fail.")
```

...but the description will no longer be true after we merge your PR.

Please go over the uses of the RF-rack-valid term in the documentation and the user messages and make sure they make sense afterwards. Because "RF-rack-valid" is a term in the already released 2025.1 version, I think we should keep its meaning.

piodul replied:
We can also consider postponing it to a follow-up if it's not a quick fix but I'd rather have this fixed in this PR.

wmitros replied:
I adjusted the uses for `RF-rack-valid` in the rebase to be consisted with the original meaning, but I believe we will want to actually change the meaning after we make it possible to fully handle keyspaces with RF < #racks by only using RF racks. Also, for that reason, I didn't create a new term for the state that `validate_rf_rack_valid_replication` is verifying

piodul commented on test/cluster/test_multidc.py:
```diff
@@ -464,3 +467,109 @@ async def test_restart_with_prefer_local(request: pytest.FixtureRequest, manager
+
+    servers = [s1, s2, s3]
+    await asyncio.gather(*[manager.api.enable_injection(s.ip_addr, "tablet_allocator_shuffle", False) for s in servers])
+    await asyncio.sleep(2)
```
:(

Is there a way to, for example, wait until at least one migration has been done - instead of relying on sleeps?

dawmd replied:
+1. Maybe we could find a way to do that in e.g. `cluster/test_tablets_migration.py`? After skimming through the file, I don't know if there's anything relevant there, but perhaps a closer look could reveal something useful

wmitros replied:
I used metrics to check whether a migration has been performed instead of sleeping in the rebase

piodul commented on service/tablet_allocator.cc:
```diff
@@ -2049,11 +2050,29 @@ class load_balancer {
 
         auto get_viable_targets = [&] () {
             std::unordered_set<host_id> viable_targets;
+            bool same_rack = false;
```
Nit: the name `same_rack` implies as this bool meant that some things are on the same rack, whereas this expresses that tablets _must be_ on the same rack. How about `must_be_on_the_same_rack`?

wmitros replied:
Makes sense, added in the rebase

piodul reviewed (commented):
Test cases look good to me.

piodul commented:
@tgrabiec or @raphaelsc - please review the changes in `service/tablet_allocator.cc`

dawmd commented on locator/tablets.hh:
```diff
@@ -716,6 +716,18 @@ class abstract_replication_strategy;
+///
+/// Because we need to investiage specific tables, the keyspace must exist.
+///
+/// This function is used should be used on startup instead of assert_rf_rack_valid_keyspace, to
+/// allow adding new racks with existing RF-rack-valid keyspaces. After a new rack is added,
+/// to create a new table in that keyspace, its RF must first be increased to the number of racks
+/// in the DC.
```
It might be better to move this explanation in `replica/database.hh` to justify calling it instead of the other function. Since it's put in `locator/tablets.hh`, I assume it's going to be used in other parts of the code where start-up might be completely orthogonal to what the check will be doing.

Here we could describe the general idea behind the check.

wmitros replied:
Makes sense, I added a similar description in `replica/database.hh` before `check_rf_rack_validity` as it fits there better

dawmd reviewed (commented):
The patch looks great! I can't comment on the load balancer logic, but the result we get (verified by the tests) is definitely something that we want.

My only doubt is whether we should somehow document the change in what operations are allowed with RF-rack-valid keyspaces and if so, where. That's not something that needs to be resolved in this PR, though.

wmitros commented on cql3/statements/create_table_statement.cc:
```diff
@@ -72,6 +72,30 @@ std::vector<column_definition> create_table_statement::get_columns() const
 
 future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector<mutation>, cql3::cql_warnings_vec>>
 create_table_statement::prepare_schema_mutations(query_processor& qp, const query_options&, api::timestamp_type ts) const {
```
Fixed in rebase

wmitros commented on locator/tablets.cc:
```diff
@@ -1101,6 +1101,85 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
     tablet_logger.debug("[assert_rf_rack_valid_keyspace]: Keyspace '{}' has been verified to be RF-rack-valid", ks);
 }
 
+void assert_rf_rack_valid_replication(const token_metadata_ptr tmptr, const replica::keyspace& ks) {
```
I initially used `assert_` to follow the approach from the function above, but I agree that `validate_` makes more sense - I renamed this in the rebase

wmitros commented on cql3/statements/create_table_statement.cc:
```diff
@@ -72,6 +72,30 @@ std::vector<column_definition> create_table_statement::get_columns() const
 
 future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector<mutation>, cql3::cql_warnings_vec>>
 create_table_statement::prepare_schema_mutations(query_processor& qp, const query_options&, api::timestamp_type ts) const {
```
Fixed in rebase

wmitros commented on locator/tablets.hh:
```diff
@@ -716,6 +716,18 @@ class abstract_replication_strategy;
+/// for all tablets in all tables in the keyspace if they are replicated on the same racks,
+/// with one replica per rack (unless the RF for the given DC is 0 or 1).
+///
+/// Because we need to investiage specific tables, the keyspace must exist.
```
Fixed in rebase

wmitros commented on locator/tablets.hh:
```diff
@@ -716,6 +716,18 @@ class abstract_replication_strategy;
+///
+/// Because we need to investiage specific tables, the keyspace must exist.
+///
+/// This function is used should be used on startup instead of assert_rf_rack_valid_keyspace, to
```
I removed this part of the comment while fixing other review

wmitros commented on locator/tablets.cc:
```diff
@@ -1101,6 +1101,85 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
+    // At this point racks_per_dc contains exactly RF racks in each DC. Now we need to make sure
+    // that all tablets in all tables in the keyspace use exactly these racks in each DC.
+    for (const auto& [cf_name, s] : ks.metadata()->cf_meta_data()) {
+        std::optional<tablet_id> curr_tablet_id = tmptr->tablets().get_tablet_map(s->id()).first_tablet();
```
Ok, added in rebase

wmitros commented:
In the rebase I fixed the review comments posted until now (mostly about naming/spelling).
I believe we want this in 2025.1.2 as that's the milestone of https://github.com/scylladb/scylladb/issues/23071, so I'm adding backport 2025.1 label

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16617/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16617/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16617/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_multidc
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4607/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7773/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15806/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16617/)
#### Build Details:
- Duration: 7 hr 47 min
- Builder: spider5.cloudius-systems.com

dawmd commented on cql3/statements/alter_keyspace_statement.cc:
```diff
@@ -285,10 +285,15 @@ cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_proce
+                //    still check that after the change the keyspace is RF-rack-valid, instead of
+                //    checking the replication again. We do that because currently we do not support
+                //    moving from one non RF-rack-valid replication strategy to another. We only
+                //    support moving from a non RF-rack-valid replication to an RF-rack-valid one.
```
It's also possible to move from one RF-rack-valid replication to a different RF-rack-valid one. For instance, consider: {DC1: 1, DC2: 0} -> {DC1: 1, DC2: 1}.

wmitros replied:
Fixed in last rebase

dawmd commented on cql3/statements/create_table_statement.cc:
```diff
@@ -72,6 +72,30 @@ std::vector<column_definition> create_table_statement::get_columns() const
 
 future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector<mutation>, cql3::cql_warnings_vec>>
 create_table_statement::prepare_schema_mutations(query_processor& qp, const query_options&, api::timestamp_type ts) const {
+    const auto tmptr = qp.proxy().get_token_metadata_ptr();
```
Note to myself: This function is executed with a Raft guard held by `schema_altering_statement`.

dawmd commented on cql3/statements/alter_keyspace_statement.cc:
```diff
@@ -276,7 +276,7 @@ cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_proce
                     locator::replication_strategy_params(ks_md_update->strategy_options(), ks_md_update->initial_tablets()));
```
nit (commit message): "removing a rack, as long as no RF-rack-valid (with RF>1) is replicated" -> "removing a rack, as long as no RF-rack-valid **keyspaces** (with RF>1) is replicated"

wmitros replied:
Fixed in last rebase

piodul commented:
> @tgrabiec or @raphaelsc - please review the changes in `service/tablet_allocator.cc`

ping (or please delegate to somebody well-versed in the tablet load balancer)

avikivity commented on cql3/statements/create_table_statement.cc:
```diff
@@ -72,6 +72,30 @@ std::vector<column_definition> create_table_statement::get_columns() const
 
 future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector<mutation>, cql3::cql_warnings_vec>>
 create_table_statement::prepare_schema_mutations(query_processor& qp, const query_options&, api::timestamp_type ts) const {
+    const auto tmptr = qp.proxy().get_token_metadata_ptr();
```
It would be better to receive it from the caller. It's an indication that we're not running a transaction with multiple conflicting metadata pointers.

wmitros replied:
`prepare_schema_mutations` is a method shared for all cql statements, maybe we can keep this as is - this will be removed in the patch that allows creating keyspaces with RF<#racks anyway

dawmd replied:
It's weird the function doesn't get a guard indeed, but for what it's worth, it's held by the caller, `schema_altering_statement::prepare_schema_mutations()`.

avikivity commented on locator/tablets.cc:
```diff
@@ -1101,6 +1101,86 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
+    const auto& nts = *static_cast<const network_topology_strategy*>(std::addressof(ars));
+
+    std::map<sstring, int> rf_per_dc;
+    bool all_dc_rf_lower_than_2 = true;
```
What's special about 2?

wmitros replied:
In the patch introducing the rf-rack-valid feature we decided (I'm not sure for which reasons) that keyspaces with RF=1 should be treated as rf-rack-valid, even though their tablets could migrate between racks. RF=0 should also be rf-rack-valid, so a bool for both of them is RF<2

avikivity replied:
I remember we wanted to treat #racks=1 as special, maybe RF=1 got confused with it.

avikivity replied:
@bhalevy suggests that it's because there are no pairing problems with RF=1. But I'd rather eliminate special cases for unimportant workloads.

avikivity commented on locator/tablets.cc:
```diff
@@ -1101,6 +1101,86 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
+    auto& first_cf_tablets = tmptr->tablets().get_tablet_map(first_s->id());
+    auto first_tablet_id = first_cf_tablets.first_tablet();
+    auto& first_tablet_info = first_cf_tablets.get_tablet_info(first_tablet_id);
+    for (auto& replica : first_tablet_info.replicas) {
```
Shouldn't you iterate over all tablets?

avikivity replied:
and all tables?

avikivity replied:
Ah, you do that later.

avikivity commented on locator/tablets.cc:
```diff
@@ -1101,6 +1101,86 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
+
+    // At this point racks_per_dc contains exactly RF racks in each DC. Now we need to make sure
+    // that all tablets in all tables in the keyspace use exactly these racks in each DC.
+    for (const auto& [cf_name, s] : ks.metadata()->cf_meta_data()) {
```
I don't think it's accurate. Each table needs to be self-consistent with its own RF.

wmitros replied:
What's your concern here? AFAIK we want all tables in a keyspace to use the same racks (and they have the same RF)

avikivity replied:
That's correct.

avikivity commented on locator/tablets.cc:
```diff
@@ -1101,6 +1101,86 @@ void assert_rf_rack_valid_keyspace(std::string_view ks, const token_metadata_ptr
+            for (const auto& [dc, racks] : curr_racks_per_dc) {
+                if (rf_per_dc[dc] < 2) {
+                    continue;
+                }
```
I don't understand the special case for < 2. I guess it's due to a previous special case with RF=1 that allows spreading over all racks. We should move towards not treating it specially.

wmitros replied:
Right, it's for that case and I'll handle the RF=1 case in the follow-up patch

wmitros commented on cql3/statements/alter_keyspace_statement.cc:
```diff
@@ -276,7 +276,7 @@ cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_proce
+                // There are three things to note here:
                 // 1. We hold a group0_guard, so it's correct to check this here.
                 //    The topology or schema cannot change while we're performing this query.
                 // 2. The replication strategy we use here does NOT represent the actual state
```
I'm preparing a separate patch that will allow creating a keyspace with RF<#racks and reducing RF. We wanted to implement this whole feature incrementally, so this patch is a small improvement in relaxing RF=#racks but it should still help with most of the test cases where we want to add a rack without that much extra complexity

wmitros commented on db/config.cc:
```diff
@@ -1397,7 +1397,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
         "Enforce RF-rack-valid keyspaces. Additionally, if there are existing RF-rack-invalid "
-        "keyspaces, attempting to start a node with this option ON will fail.")
+        "keyspaces, attempting to start a node with this option ON will fail unless they became RF-rack-invalid"
+        "due to adding new racks without increasing their RF.")
```
I'm planning to redefine it in the mentioned patch that's in progress. I didn't redefine it in this patch because of the RF=1 case which still doesn't use exactly one rack

avikivity replied:
I feel like much could be simplified if we could get rid of the RF=1 special case.

Where is it in the requirements?

piodul commented:
@wmitros why "Refs: https://github.com/scylladb/scylladb/issues/23426", and not Fixes? is there something left to be done for this particular issue?

wmitros commented:
> @wmitros why "Refs: #23426", and not Fixes? is there something left to be done for this particular issue?

I used only "Refs" because #23426 suggests that a rf-rack-valid keyspace should be bound to specific racks but this is not fully implemented here - we add a limitation that prevents creating new tables in it if if there are other racks.
But that's probably a small concern - I'll change it to a "Fixes" in the rebase

wmitros commented:
In the new version:
* fixed commit message
* fixed comments
* in preparation for allowing RF<#racks, the new check is more strict - we check that all replicas of all tablets in a keyspace have the same rack order
* added a check from `assert_rf_rack_valid_keyspace` for confirming that the keyspace's replication strategy knows all racks from the provided token_metadata

wmitros commented:
Looks like the replica rack order restriction was premature, I removed it after all

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16745/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16745/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16745/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/test_multidc
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15967/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7923/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16745/)
#### Failed Tests (2/43692):
- [test_base_partition_deletion_with_metrics[True].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16745/testReport/junit/cqlpy/test_materialized_view/Tests___Unit_Tests___test_base_partition_deletion_with_metrics_True__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_base_partition_deletion_with_metrics[True].debug.3)
- [test_enable_rbno_extra_nodes_multi_dc_rack](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16745/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest_with_consistent_topology_changes___test_enable_rbno_extra_nodes_multi_dc_rack) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_enable_rbno_extra_nodes_multi_dc_rack)
#### Build Details:
- Duration: 8 hr 2 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/23559)

**Changed Files:**
- [operating-scylla/procedures/config-change/index.html](http://previews.docs.scylladb.com/scylladb/23559/operating-scylla/procedures/config-change/index.html)
- [operating-scylla/procedures/config-change/sstable-dictionary-compression](http://previews.docs.scylladb.com/scylladb/23559/operating-scylla/procedures/config-change/sstable-dictionary-compression)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16774/)
:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16774/)
#### Build Failure:
```
[2025-04-15T08:47:37.137Z] FAILED: build/debug/locator/tablets.o 
[2025-04-15T08:47:37.137Z] clang++ -MD -MT build/debug/locator/tablets.o -MF build/debug/locator/tablets.o.d -std=gnu++23 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/gen/include -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=19 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEBUG_PROMISE -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_THREAD_NO_LIB -DBOOST_THREAD_DYN_LINK -DFMT_SHARED -I/usr/include/p11-kit-1 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -DSCYLLA_BUILD_MODE=debug -g -gz -Xclang -fexperimental-assignment-tracking=disabled -iquote. -iquote build/debug/gen -std=gnu++23 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -Ikmipc/kmipc-2.1.0t-rhel84_64/include -DHAVE_KMIP -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN  -c -o build/debug/locator/tablets.o locator/tablets.cc
[2025-04-15T08:47:37.138Z] locator/tablets.cc:1165:17: error: use of undeclared identifier rf_per_dc
[2025-04-15T08:47:37.138Z]  1165 |             if (rf_per_dc[dc] < 2) {
[2025-04-15T08:47:37.138Z]       |                 ^
[2025-04-15T08:47:37.138Z] 1 error generated.
```
#### Build Details:
- Duration: 27 min
- Builder: spider5.cloudius-systems.com

mykaul commented:
@wmitros , @piodul - please push this forward, in the hope it'll help stabilize CI!

Deexie commented:
What's the status of this and https://github.com/scylladb/scylladb/pull/23738 PRs?

wmitros commented:
> What's the status of this and #23738 PRs?

They will be mostly replaced by the work @tgrabiec and @bhalevy are doing. So I should better close these 2 PRs
